### ISSUE TITLE
⌨ Accessibility Issue: Unsaved Changes Interruptions

### ISSUE THREAD

**Issue by RM-1978** [GCTC-NTGC/gc-digital-talent#5359]
## ❌ Current behaviour (issue)
The screen reader keeps flagging that a user has ‘unsaved changes’ each time they add information in a new field for every form they encounter. 

### How to reproduce
**AT configuration:** JAWS/NVDA Screen Reader, Windows 10 & 11, Chrome/Edge 

1.	Open JAWS/NVDA in Windows in Chrome or Edge.  
2.	Go to the About Me component of the Application profile.
3.	Using MS Narrator or Voice Over, enter text in any form field. 
4.	Navigate using your keyboard and fill in text field information while listening to the programmatic feedback to the screen reader: repeated “unsaved changes” statement. 
**Screen Capture** 
## ✔ Expected behaviour (fix)
Users indicate they only need to be advised that they have unsaved information once per page. Notify the user once at the end of the page with a list missing entries or unsaved changes. Perhaps, listing unsaved changes as they go through the UI in a prominent container somewhere on the page before they encounter the Save and Go Back button?

### Notes
Providing an error notification in the page title and an alert dialog box at the top of the page identifying the mandatory fields which may have been omitted or unsaved would improve their experience. 

**Potential WCAG Considerations (if any)**  
[4.1.3 Status Messages](https://w3c.github.io/wcag/understanding/status-messages),  [3.3.1 Error Identification ](https://www.w3.org/WAI/WCAG22/Understanding/error-identification) 

### Acceptance Criteria
A set of assumptions which, when tested, verify that the accessibility requirement was met.
- [ ] Criteria
- [ ] Criteria 2

**Comment by petertgiles** [comment]
Thanks for this issue, Rosalie!  I'm trying to figure out the acceptance criteria for this one but it's a little unclear to me.  Is the issue to:
1. Make sure screen reader only says "unsaved changes" once
2. Bring back the unsaved changes component at the top of the page (see #4980)
3. Add "unsaved changes" to the page title
Or some combination of these?

**Comment by RM-1978** [comment]
Hi Peter, thanks for the question!  I think the issue is the same  one flagged in  #4980 (see https://github.com/GCTC-NTGC/gc-digital-talent/issues/4980), with one addition: Make sure screen reader only says "unsaved changes" once when reloaded.

**Comment by petertgiles** [comment]
Thanks, I've updated the acceptance criteria.  Feel free to tweak if it's not quite right yet.

**Comment by RM-1978** [comment]
It is perfect just the way it is. Thanks so much for taking the time to collaborate on this!

**Comment by petertgiles** [comment]
Some questions while trying to estimate:
- When should it be read out? (beginning, end)
- We will have problems with the page shifting again?
- Should we abandon the unsaved change strategy?

**Comment by esizer** [comment]
> I'm confused by the general description of what's happening in the original issue (The Fable video is restricted, so I can't watch it). In an ideal world, I would have suggested that the "Unsaved changes" text only be read to users if they return focus to a field they've already changed (e.g. "input label => required => unsaved changes if applicable"). This means that the text would only be read on re-focus of an applicable input.

@substrae This is essentially what is happening

> I think it is important to see how annoyed the user gets with the interruption. I think it is important for us to look at a solution because this irritant was repeated with other users in their self tests and interviews.

@RM-1978 I haven't seen the video yet, but I already understand their frustrations. I was not a fan of the introduction of this feature in the first place. I feel as though if a user changes a field, they know they did and being told every time could get annoying quick especially when it is being announced ahead of more important information 😕 

---

The main reason we added this feature was because during user testing, French speakers voiced concerns over the translations and actually felt the English (their second language) was better so they would switch languages and they would lose their unsaved changes. We needed some way to communicate that with the user and used an alert at the top of the page. However, we got more feedback indicating this was not expected and it caused layout shift that annoyed users.

We moved to use `aria-describedby` to communicate it to users which is the cause of it being announced on any field that is changed as a way to communicate to a user that field has changed is not saved.

I'm not sure what the best answer is here outside of trying to improve our French version of forms so French speakers don't feel the need to use their second language to understand the labels 😢 This whole feature is duct tape to fix a core issue of having poor translations.

I would be happy to remove the feature entirely and see what we can do about fixing the French form translations.

**Comment by substrae** [comment]
Hmm, highlighted unsaved changes isn't an unusual feature @esizer. It's not just beneficial to people who switch languages, it's also helpful for people who leave and return mid-task, as well as people who accidentally navigate away from the page without saving.

>Irritating notifications while entering information: While entering information into the form, I continue to receive messages saying that I either have an error in the form or that I have not saved my updated information. **Sometimes these notifications continue to display after each key that has been pressed.**

It doesn't sound like what I described as an ideal is actually happening here - I'm suggesting that if a user changes text, they receive no notification whatsoever unless they navigate away from that input and then back to it.

**Comment by esizer** [comment]
> Hmm, highlighted unsaved changes isn't an unusual feature @esizer. It's not just beneficial to people who switch languages, it's also helpful for people who leave and return mid-task, as well as people who accidentally navigate away from the page without saving.

I agree having it save your changes for these scenarios is a good change. I'm more not a fan of how I implemented it from a code stand point 😜 I don't think I was very clear on that. I just think we may have been a little over eager in implementing based on all of the feedback/changes we have been getting surrounding it. 😅 

> After watching the video, it does seem he's referring to unsaved changes being identified at the input-level. So I think this comes down to whether or not identifying which field is changed warrants the annoyance here. Alternatively, we separate the behaviour for auditory users - provide them with an auditory summary when they enter the form itself and then don't repeat unsaved changes again after that.

Maybe putting an `aria-hidden` on the unsaved changes message? That should separate the auditory and visual parts. I think the most valuable is the alert at the top of the page (which seems to be what the user indicated in the video but can't be certain on this). 

Also, I wouldn't mind having the errors and unsaved changes show on `onBlur` instead of `onChange` as well but that is probably a different discussion. I just couldn't get that working initially while working on it.

**Comment by substrae** [comment]
Ahhh okay, I see what you're saying. Couple things then:
- I think we're generally guilty of making changes to things before verifying if it's a problem at scale - so this feature is no different.
- I think we should add a summary above the form for auditory users only (`visually-hidden` should be helpful here).
- In the same vein, I agree we can probably use `aria-hidden` on the visual label portion as well.
- I think from an accessibility standpoint changing things to `onBlur` will make things WAY more tolerable. Maybe we can talk through the nuances there just to be sure.

**Comment by esizer** [comment]
> * I think we're generally guilty of making changes to things before verifying if it's a problem at scale - so this feature is no different.
> * I think we should add a summary above the form for auditory users only (`visually-hidden` should be helpful here).
> * In the same vein, I agree we can probably use `aria-hidden` on the visual label portion as well.
> * I think from an accessibility standpoint changing things to `onBlur` will make things WAY more tolerable. Maybe we can talk through the nuances there just to be sure.

💯 on all points

**Comment by esizer** [comment]
We no longer display "Unsaved changes" and instead just highlight the field as of https://github.com/GCTC-NTGC/gc-digital-talent/pull/6895. Are we safe to close this?

**Comment by esizer** [comment]
Closing since we no longer display unsaved change messages
